packages/api-server/src/convert-tx.ts

```diff
@@ -132,45 +132,41 @@ export function getSignature(tx: PolyjuiceTransaction): HexString {
   return "0x" + tx.r.slice(2) + tx.s.slice(2) + realVWithoutPrefix;
 }
 
-function numberToRlpEncode(num: HexString): HexString {
-  if (num === "0x0" || num === "0x") {
-    return "0x";
-  }
-
-  return "0x" + BigInt(num).toString(16);
-}
-
+// https://eips.ethereum.org/EIPS/eip-155
+// For non eip-155 txs, (nonce, gasprice, startgas, to, value, data)
+// For eip155 txs, (nonce, gasprice, startgas, to, value, data, chainid, 0, 0)
 function calcMessage(tx: PolyjuiceTransaction): HexString {
-  let vInt = +tx.v;
-  let finalVInt = undefined;
+  const v: bigint = BigInt(tx.v);
+
+  const beforeEncode: any[] = [
+    toRlpNumber(tx.nonce),
+    toRlpNumber(tx.gasPrice),
+    toRlpNumber(tx.gasLimit),
+    tx.to,
+    toRlpNumber(tx.value),
+    tx.data,
+  ];
 
-  if (vInt === 27 || vInt === 28) {
-    throw new Error(
-      `only EIP155 transaction is allowed, illegal transaction recId ${tx.v}. more info: ${COMPATIBLE_DOCS_URL}`
-    );
-  }
+  // if v = 27 / 28, it's non eip-155 txs
+  if (v !== 27n && v !== 28n) {
+    let chainId: bigint;
+    if (v % 2n === 0n) {
+      chainId = (v - 36n) / 2n;
+    } else {
+      chainId = (v - 35n) / 2n;
+    }
 
-  if (vInt % 2 === 0) {
-    finalVInt = "0x" + BigInt((vInt - 36) / 2).toString(16);
-  } else {
-    finalVInt = "0x" + BigInt((vInt - 35) / 2).toString(16);
+    // chain id
+    beforeEncode.push(chainId);
+    // r
+    beforeEncode.push(0);
+    // s
+    beforeEncode.push(0);
   }
 
-  const rawTx: PolyjuiceTransaction = {
-    ...tx,
-    nonce: numberToRlpEncode(tx.nonce),
-    gasPrice: numberToRlpEncode(tx.gasPrice),
-    gasLimit: numberToRlpEncode(tx.gasLimit),
-    value: numberToRlpEncode(tx.value),
-    r: "0x",
-    s: "0x",
-    v: numberToRlpEncode(finalVInt),
-  };
-
-  const encoded = encodePolyjuiceTransaction(rawTx);
+  const encoded: Buffer = rlp.encode(beforeEncode);
```

Comment:
Creating standalone util functions to make it more readable would be helpful, such as `bool isEip155Transaction`, `bytes rlpEncodeEip155Transaction`, and `bytes rlpEncodeNonEip155Transaction`.